### PR TITLE
Add golangci-lint

### PR DESCRIPTION
**Pull request by chris124567**
I have been adding this to our repos so we can do things like check struct tags automatically.  I know core is still changing so if this is unwanted right now please feel free to close this PR.

Edit: Need to do some Ledger review fixes today.  Will fix remaining warnings tomorrow.

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -71,7 +71,7 @@ type Network struct {
 		Height    uint64        `json:"height"`
 		OakTime   time.Duration `json:"oakTime"`
 		OakTarget types.BlockID `json:"oakTarget"`
-	} `json:"hardforkASIC"`
+	} `json:"hardforkAsic"`
```
Comment (lukechampine) [inline consensus/state.go:74]: definitely not. ASIC is an acronym.

**Comment by lukechampine** [comment]
Let's disable `modifies-value-receiver` and `indent-error-flow`. `if-return` is stupid too (it makes refactoring more annoying) but I guess it's situational, and it only comes up in `smux` which I don't care about.

File: consensus/state.go [hunk 2]
```diff
@@ -67,6 +67,7 @@ type Network struct {
 		FixHeight        uint64    `json:"fixHeight"`
 		GenesisTimestamp time.Time `json:"genesisTimestamp"`
 	} `json:"hardforkOak"`
+	//nolint:tagliatelle
```
Comment (lukechampine) [inline consensus/state.go:70]: unfortunate. There's an issue open for specifying custom initialisms in tagliatelle, but no indication that it'll be implemented any time soon.

Comment (lukechampine) [reply consensus/state.go:70]: btw, old tag name needs to be restored

Comment (chris124567) [reply consensus/state.go:70]: Sorry about that... Didn't sleep well last night

**Comment by n8maninger** [comment]
IIRC, we fixed the CI fail on Windows by adding the following step:
```yml
- name: Configure git # required for golangci-lint on Windows
  run: git config --global core.autocrlf false
```

**Pull request merged**